tests/unit/stream_alert_rule_processor/test_alert_forward.py

```diff
@@ -0,0 +1,209 @@
+        ])
+
+    @staticmethod
+    def _generate_alerts(count):
+        """Generate a list of alert dictionaries."""
+        return [
+            {
+                'record': '{"abc": 123}',
+                'rule_description': 'Desc{}'.format(i),
+                'rule_name': 'Rule{}'.format(i),
+                'outputs': ['aws-lambda:...', 'aws-s3:...']
+            }
+            for i in xrange(count)
+        ]
+
+    def test_alert_batches_single_alert(self):
+        """AlertForwarder - Alert Batching - Single Alert"""
+        alerts = self._generate_alerts(1)
+        result = list(self.forwarder._alert_batches(alerts))
+
+        expected = [
+            {
+                self.ALERT_TABLE: [
+                    {
+                        'PutRequest': {
+                            'Item': {
+                                'RuleName': {'S': 'Rule0'},
+                                'Timestamp': {'S': ANY},
+                                'Cluster': {'S': 'corp'},
+                                'RuleDescription': {'S': 'Desc0'},
+                                'Outputs': {'SS': ['aws-lambda:...', 'aws-s3:...']},
+                                'Record': {'S': '"{\\"abc\\": 123}"'},
+                                'TTL': {'N': ANY}
+                            }
+                        }
+                    }
+                ]
+            }
+        ]
+        assert_equal(expected, result)
+
+    def test_alert_batches_max_batch(self):
+        """AlertForwarder - Alert Batching - Full Batch"""
+        alerts = self._generate_alerts(10)
+        result = list(self.forwarder._alert_batches(alerts, batch_size=10))
+
+        assert_equal(1, len(result))  # There should be one batch.
+        assert_equal(10, len(result[0][self.ALERT_TABLE]))  # All 10 alerts should be in the batch.
+
+    def test_alert_batches_max_batch_plus_one(self):
+        """AlertForwarder - Alert Batching - Full Batch + 1"""
+        alerts = self._generate_alerts(11)
+        result = list(self.forwarder._alert_batches(alerts, batch_size=10))
+
+        assert_equal(2, len(result))  # There should be 2 alert batches.
+        assert_equal(10, len(result[0][self.ALERT_TABLE]))  # 10 alerts in the first batch.
+        assert_equal(1, len(result[1][self.ALERT_TABLE]))  # 1 alert in the second batch.
+
+    @patch('stream_alert.rule_processor.alert_forward.LOGGER')
+    def test_dynamo_unprocessed_alerts(self, log_mock):
+        """AlertForwarder - Dynamo - Retry unprocessed alerts"""
+
+        def mock_batch_write_item(**kwargs):
+            """Mock client_dynamo.batch_write_item to always return all items unprocessed."""
+            return {
+                'ResponseMetadata': {'HTTPStatusCode': 200},
+                'UnprocessedItems': {self.ALERT_TABLE: kwargs['RequestItems']}
+            }
+
+        self.boto_mock.return_value.batch_write_item.side_effect = mock_batch_write_item
```

Comment:
RE your previous (now not shown) comment about this needing to be dynamic - the value of `kwargs['RequestItems']` is always going to be the same thing as what is returned from the `self._generate_alerts(1)` function that is passed in (in this case). IE - this is effectively the same:

```
        ...
        items = self._generate_alerts(1)
        def mock_batch_write_item(**kwargs):
            """Mock client_dynamo.batch_write_item to always return all items unprocessed."""
            return {
                'ResponseMetadata': {'HTTPStatusCode': 200},
                'UnprocessedItems': {self.ALERT_TABLE: items}
            }

        self.boto_mock.return_value.batch_write_item.side_effect = mock_batch_write_item

        self.forwarder._send_to_dynamo(items)
        ...
```

Reply:
It is not anymore - after adding the metrics I updated the test so that it invoked multiple times with different batch sizes